firedrake/slate/preconditioners.py

```diff
@@ -0,0 +1,285 @@
+        self._assemble_S = create_assembly_callable(
+            schur_comp,
+            tensor=self.S,
+            bcs=trace_conditions,
+            form_compiler_parameters=context.fc_params)
+
+        self._assemble_S()
+        self.S.force_evaluation()
+        Smat = self.S.petscmat
+
+        # Nullspace for the multiplier problem
+        nullsp = P.getNullSpace()
+        if nullsp.handle != 0:
+            new_vecs = get_trace_nullspace_vecs(K * Atilde.inv, nullsp,
+                                                V, V_d, TraceSpace)
+            tr_nullsp = PETSc.NullSpace().create(vectors=new_vecs,
+                                                 comm=pc.comm)
+            Smat.setNullSpace(tr_nullsp)
+
+        # Set up the KSP for the system of Lagrange multipliers
+        ksp = PETSc.KSP().create(comm=pc.comm)
+        ksp.setOptionsPrefix(prefix + "trace_")
+        ksp.setTolerances(rtol=1e-13)
+        ksp.setOperators(Smat)
+        ksp.setUp()
+        ksp.setFromOptions()
+        self.ksp = ksp
+
+        # Now we construct the local tensors for the reconstruction stage
+        # TODO: Add support for mixed tensors and these variables
+        # become unnecessary
+        split_forms = split_form(new_form)
+        A = Tensor(next(sf.form for sf in split_forms
+                        if sf.indices == (0, 0)))
+        B = Tensor(next(sf.form for sf in split_forms
+                        if sf.indices == (1, 0)))
+        C = Tensor(next(sf.form for sf in split_forms
+                        if sf.indices == (1, 1)))
+        trial = TrialFunction(FunctionSpace(V.mesh(),
+                                            BrokenElement(W.ufl_element())))
+        K_local = Tensor(gammar('+') * ufl.dot(trial, n) * ufl.dS)
+
+        # Split functions and reconstruct each bit separately
+        sigma_h, u_h = self.broken_solution.split()
+        g, f = self.broken_rhs.split()
+
+        # Pressure reconstruction
+        M = B * A.inv * B.T + C
+        u_sol = M.inv * f + M.inv * (B * A.inv *
+                                     K_local.T * self.trace_solution
+                                     - B * A.inv * g)
+        self._assemble_pressure = create_assembly_callable(
+            u_sol,
+            tensor=u_h,
+            form_compiler_parameters=context.fc_params)
+
+        # Velocity reconstruction
+        sigma_sol = A.inv * g + A.inv * (B.T * u_h -
+                                         K_local.T * self.trace_solution)
+        self._assemble_velocity = create_assembly_callable(
+            sigma_sol,
+            tensor=sigma_h,
+            form_compiler_parameters=context.fc_params)
+
+        # Set up the projector for projecting the broken solution
+        # into the unbroken finite element spaces
+        # NOTE: Tolerance here matters!
+        sigma_b, _ = self.broken_solution.split()
+        sigma_u, _ = self.unbroken_solution.split()
```

Comment:
Implicit assumption of `V*P`.  Also, continuing inconsistency in what you call these things.